service/sharddistributor/leader/process/processor.go

```diff
@@ -160,6 +177,56 @@ func (p *namespaceProcessor) runProcess(ctx context.Context) {
 	}
 }
 
+// runCleanupLoop periodically removes stale executors.
+func (p *namespaceProcessor) runCleanupLoop(ctx context.Context) {
+	ticker := p.timeSource.NewTicker(p.cfg.HeartbeatTTL)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.logger.Info("Cleanup loop cancelled.")
+			return
+		case <-ticker.Chan():
+			p.logger.Info("Periodic heartbeat cleanup triggered.")
+			p.cleanupStaleExecutors(ctx)
+		}
+	}
+}
+
+// cleanupStaleExecutors removes executors who have not reported a heartbeat recently.
+func (p *namespaceProcessor) cleanupStaleExecutors(ctx context.Context) {
+	// 1. Get the current heartbeat states. We don't need assignments for this operation.
+	heartbeatStates, _, _, err := p.shardStore.GetState(ctx)
+	if err != nil {
+		p.logger.Error("Failed to get state for heartbeat cleanup", tag.Error(err))
+		return
+	}
+
+	// 2. Identify expired executors.
+	var expiredExecutors []string
+	now := p.timeSource.Now().Unix()
+	heartbeatTTL := int64(p.cfg.HeartbeatTTL.Seconds())
+
+	for executorID, state := range heartbeatStates {
+		if (now - state.LastHeartbeat) > heartbeatTTL {
```

Comment:
I think it's fine, but I would prefer using the time library here - calculating directly on time is always scary

Reply:
Well, state.LastHeartbeat is in UNIX timestamp, so I don't think it could go wrong if we operate on UNIX timestamps.